### PR TITLE
Use remote id

### PR DESCRIPTION
**Pull request by mario**
Initially intended to be a cherrypick of oC PR 2138, had to change it since our code is different.

Signed-off-by: Mario Danic <mario@lovelyhq.com>

### CODE REVIEW

**Comment by AndyScherzinger** [comment]
:+1: fine by me :)

**Comment by tobiasKaminsky** [comment]
This needs an intensive test with E2E, as sync is done with E2E via metadata and the filename is also set via metadata.

**Comment by mario** [comment]
@tobiasKaminsky do you want to wait until 3.2 with this then?

**Comment by AndyScherzinger** [comment]
@mario I created a stable-3.1.x branch so this PR can be merged when approved by you and the conflicts have been resolved.

**Comment by mario** [comment]
@AndyScherzinger I'd say @tobiasKaminsky needs to test this.

File: src/main/java/com/owncloud/android/datamodel/FileDataStorageManager.java [hunk 1]
```diff
@@ -120,6 +120,26 @@ public OCFile getFileByPath(String path) {
...
 
+    /**
+     * This will return a OCFile by its given FileId here refered as the remoteId.
+     * Its the fileId ownCloud Core uses to identify a file even if its name has changed.
+     *
+     *
+     * @param remoteID
+     * @return
+     */
+    public OCFile getFileByRemoteId(String remoteID) {
```
Comment (tobiasKaminsky) [inline src/main/java/com/owncloud/android/datamodel/FileDataStorageManager.java]: This function is never used?

File: src/main/java/com/owncloud/android/operations/SynchronizeFolderOperation.java [hunk 2]
```diff
@@ -284,7 +284,7 @@ private void synchronizeData(ArrayList<Object> folderAndFiles) throws OperationC
         List<OCFile> localFiles = storageManager.getFolderContent(mLocalFolder, false);
         Map<String, OCFile> localFilesMap = new HashMap<>(localFiles.size());
         for (OCFile file : localFiles) {
-            localFilesMap.put(file.getRemotePath(), file);
+            localFilesMap.put(file.getRemoteId(), file);
```
Comment (tobiasKaminsky) [inline src/main/java/com/owncloud/android/operations/SynchronizeFolderOperation.java]: Here remoteId is used as key, but on https://github.com/nextcloud/android/pull/2375/files#diff-a113e46c3018f8daf5d37fbcc3ec5739L302
it is still used remotePath.

**Comment by tobiasKaminsky** [comment]
@mario, I commented on it.

File: src/main/java/com/owncloud/android/datamodel/FileDataStorageManager.java (hunk 1 above)
Comment (mario) [reply src/main/java/com/owncloud/android/datamodel/FileDataStorageManager.java]: Indeed. But it might be used later on.

**Comment by mario** [comment]
PR updated.

File: src/main/java/com/owncloud/android/datamodel/FileDataStorageManager.java (hunk 1 above)
Comment (tobiasKaminsky) [reply src/main/java/com/owncloud/android/datamodel/FileDataStorageManager.java]: Then we can implement it later on anyways, so please remove.

File: src/main/java/com/owncloud/android/operations/SynchronizeFolderOperation.java [hunk 3]
```diff
@@ -299,7 +299,7 @@ private void synchronizeData(ArrayList<Object> folderAndFiles) throws OperationC
 
             /// retrieve local data for the read file
             //  localFile = mStorageManager.getFileByPath(remoteFile.getRemotePath());
-            localFile = localFilesMap.remove(remoteFile.getRemotePath());
+            localFile = localFilesMap.remove(remoteFile.getRemoteId());
```
Comment (tobiasKaminsky) [inline src/main/java/com/owncloud/android/operations/SynchronizeFolderOperation.java]: As this is only a key for the hashmap, this change is not affecting anything at all, or?

File: src/main/java/com/owncloud/android/operations/SynchronizeFolderOperation.java [hunk 4]
```diff
@@ -330,6 +330,7 @@ private void updateLocalStateData(OCFile remoteFile, OCFile localFile, OCFile up
             updatedFile.setModificationTimestampAtLastSyncForData(
                     localFile.getModificationTimestampAtLastSyncForData()
             );
+            updatedFile.setFileName(remoteFile.getFileName());
```
Comment (tobiasKaminsky) [inline src/main/java/com/owncloud/android/operations/SynchronizeFolderOperation.java]: So it boils down to this one, or?

**Comment by tobiasKaminsky** [comment]
- sync a folder -> all files in it are downloaded
- rename a file in folder via webUI
- sync same folder again

```
04-27 12:01:47.035 4241-4276/com.nextcloud.client E/AndroidRuntime: FATAL EXCEPTION: Syncfolder thread
    Process: com.nextcloud.client, PID: 4241
    java.lang.IllegalArgumentException: serverFile and localFile do not correspond to the same OC file
        at com.owncloud.android.operations.SynchronizeFileOperation.<init>(SynchronizeFileOperation.java:125)
        at com.owncloud.android.operations.SynchronizeFolderOperation.classifyFileForLaterSyncOrDownload(SynchronizeFolderOperation.java:369)
        at com.owncloud.android.operations.SynchronizeFolderOperation.synchronizeData(SynchronizeFolderOperation.java:315)
        at com.owncloud.android.operations.SynchronizeFolderOperation.fetchAndSyncRemoteFolder(SynchronizeFolderOperation.java:225)
        at com.owncloud.android.operations.SynchronizeFolderOperation.run(SynchronizeFolderOperation.java:150)
        at com.owncloud.android.lib.common.operations.RemoteOperation.execute(RemoteOperation.java:145)
        at com.owncloud.android.lib.common.operations.RemoteOperation.execute(RemoteOperation.java:149)
        at com.owncloud.android.operations.common.SyncOperation.execute(SyncOperation.java:93)
        at com.owncloud.android.services.SyncFolderHandler.doOperation(SyncFolderHandler.java:122)
        at com.owncloud.android.services.SyncFolderHandler.handleMessage(SyncFolderHandler.java:91)
        at android.os.Handler.dispatchMessage(Handler.java:106)
        at android.os.Looper.loop(Looper.java:164)
        at android.os.HandlerThread.run(HandlerThread.java:65)
``` 

This does not happen on current master.

**Comment by mario** [comment]
Uh. Strange. Will see but not in the near future as this is a cherrypick from the dark side.

File: src/main/java/com/owncloud/android/operations/SynchronizeFolderOperation.java (hunk 4 above)
Comment (tobiasKaminsky) [reply src/main/java/com/owncloud/android/operations/SynchronizeFolderOperation.java]: We have this is already in 

public static OCFile fillOCFile(RemoteFile remote) {
        OCFile file = new OCFile(remote.getRemotePath());
…
}

**Comment by tobiasKaminsky** [comment]
As the only change is already included in our "fillOCFile", I am closing this.

**Pull request closed without merging**